cli/src/cmd/forge/verify/sourcify.rs

```diff
@@ -0,0 +1,166 @@
+use std::{collections::HashMap, fs, path::PathBuf};
+
+use async_trait::async_trait;
+use cast::SimpleCast;
+use ethers::solc::artifacts::output_selection::ContractOutputSelection;
+use foundry_utils::Retry;
+use futures::FutureExt;
+use serde::{Deserialize, Serialize};
+use tracing::{trace, warn};
+
+use crate::cmd::LoadConfig;
+
+use super::{VerificationProvider, VerifyArgs, VerifyCheckArgs};
+
+pub static SOURCIFY_URL: &'static str = "https://sourcify.dev/server/";
+
+pub type SourcifyVerificationResponse = Vec<SourcifyResponseElement>;
+
+#[derive(Deserialize, Debug)]
+pub struct SourcifyResponseElement {
+    status: String,
+    #[serde(rename = "storageTimestamp")]
+    storage_timestamp: Option<String>,
+}
+
+#[derive(Serialize, Debug)]
+pub struct SourcifyVerifyRequest {
+    address: String,
+    chain: String,
+    files: HashMap<String, String>,
+    #[serde(rename = "chosenContract", skip_serializing_if = "Option::is_none")]
+    chosen_contract: Option<String>,
+}
+
+pub struct SourcifyVerificationProvider {}
+
+#[async_trait]
+impl VerificationProvider for SourcifyVerificationProvider {
+    async fn verify(&self, args: VerifyArgs) -> eyre::Result<()> {
+        let config = args.load_config_emit_warnings();
+        let project = config.project()?;
+
+        if !config.extra_output_files.contains(&ContractOutputSelection::Metadata) {
+            eyre::bail!("Metadata is required for sourcify verification. Try adding `extra_output_files = [\"metadata\"]` to `foundry.toml`")
+        }
+
+        let path = match args.contract.path {
+            Some(path) => PathBuf::from(path),
+            None if config.cache => {
+                let cache = project.read_cache_file()?;
+                crate::cmd::get_cached_entry_by_name(&cache, &args.contract.name)?.0
+            }
+            _ => eyre::bail!(
+                "If cache is disabled, contract info must be provided in the format <path>:<name>"
+            ),
+        };
```

Comment:
for separate PR but this "read artifact from name" logic should go behind a helper fn